### ISSUE TITLE
Error running map_cells_to_space

### ISSUE THREAD

**Issue by smorabit** [broadinstitute/Tangram#43]
Hi, I am interested in using Tangram for my research. I have Visium data and 10X genomics snRNA-seq data. I am following the [tutorial](https://github.com/broadinstitute/Tangram/blob/master/tangram_tutorial.ipynb), which works flawlessly for the data provided, but gives an error when running `map_cells_to_space` on my own data. 

I am loading the Visium data straight from spaceranger, and I am constructing an anndata object from a `.mtx` file for the single nucleus data. 

Here is the code that I am running:
```
import os, sys
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import scanpy as sc
import torch
from scipy import io
from scipy.sparse import coo_matrix, csr_matrix
import anndata

#sys.path.append('/home/exx/git/Tangram/')  # uncomment for local import
import tangram as tg

%load_ext autoreload
%autoreload 2
%matplotlib inline

data_dir = 'data/'

# load one visium sample:
ad_sp = sc.read_visium('Visium1/outs/')

# load single-cell data
X = io.mmread("{}zhou_counts.mtx".format(data_dir))

# create anndata object
ad_sc = anndata.AnnData(
    X=X.transpose().tocsr()
)

# load sample metadata:
sample_meta = pd.read_csv("{}zhou_meta.csv".format(data_dir))

# load gene names:
with open("{}zhou_gene_names.csv".format(data_dir), 'r') as f:
    gene_names = f.read().splitlines()
    
ad_sc.obs = sample_meta
ad_sc.obs.index = ad_sc.obs['barcode']
ad_sc.obs = ad_sc.obs.drop(labels='barcode', axis=1)
ad_sc.var.index = gene_names

markers = pd.read_csv('data/zhou_marker_genes.csv')

# only keep markers that are in both dataset:
markers = markers[markers.gene.isin(ad_sp.var.index)]

# prepare for mapping
tg.pp_adatas(ad_sc, ad_sp, genes=markers.gene.unique())

assert ad_sc.uns['training_genes'] == ad_sp.uns['training_genes']

ad_map = tg.map_cells_to_space(
    adata_sc=ad_sc,
    adata_sp=ad_sp,
    #device='cpu'
    device='cuda'
)
```

The error when running `map_cells_to_space` is the following:

```
INFO:root:Allocate tensors for mapping.
INFO:root:Begin training with 1500 genes and rna_count_based density_prior in cells mode...
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-130-a433d3224cd2> in <module>
      3     adata_sp=ad_sp,
      4     #device='cpu'
----> 5     device='cuda'
      6 )

/dfs3b/swaruplab/smorabit/bin/software/miniconda3/envs/scvi-env/lib/python3.7/site-packages/tangram/mapping_utils.py in map_cells_to_space(adata_sc, adata_sp, cv_train_genes, cluster_label, mode, device, learning_rate, num_epochs, scale, lambda_d, lambda_g1, lambda_g2, lambda_r, lambda_count, lambda_f_reg, target_count, random_state, verbose, density_prior)
    311         )
    312         mapper = mo.Mapper(
--> 313             S=S, G=G, d=d, device=device, random_state=random_state, **hyperparameters,
    314         )
    315 

/dfs3b/swaruplab/smorabit/bin/software/miniconda3/envs/scvi-env/lib/python3.7/site-packages/tangram/mapping_optimizer.py in __init__(self, S, G, d, d_source, lambda_g1, lambda_d, lambda_g2, lambda_r, device, adata_map, random_state)
     59         self.target_density_enabled = d is not None
     60         if self.target_density_enabled:
---> 61             self.d = torch.tensor(d, device=device, dtype=torch.float32)
     62 
     63         self.source_density_enabled = d_source is not None

ValueError: too many dimensions 'matrix'
```

**Comment by Chengwei94** [comment]
I got the same error. tg.pp_adatas changed the shape of my matrix.

**Comment by ziqlu0722** [comment]
Hi @smorabit  and @Chengwei94, thanks for your interest in using Tangram! To reproduce the issue, could you please share a little more details? For example, how did you set up your environment, is it possible to share a dummy/demo dataset with the similar format of data you've been working with? Thank you!

**Comment by Chengwei94** [comment]
Hi @ziqlu0722, 


I believe that I found the error lies in tg.pp_adata function that causes my adata_sp.obs to become corrupted with this error. The truth value of an array with more than one element is ambiguous. Use a.any() or a.all().

Once I changed the code from      

``
rna_count_per_spot = adata_sp.X.sum(axis=1)
adata_sp.obs["rna_count_based_density"] = rna_count_per_spot / np.sum(
        rna_count_per_spot
    )
``

to 

``
rna_count_per_spot = adata_sp.X.sum(axis=1)
adata_sp.obs["rna_count_based_density"] = pd.DataFrame(rna_count_per_spot / np.sum(
        rna_count_per_spot), index = adata_sp.obs.index
),
``

the code worked for me.

Cheers

**Comment by lewlin** [comment]
@Chengwei94 and @smorabit could you give us more info? We can't reproduce your error. In particular we need: all libraries versions (pytorch, CUDA, scanpy, squidpy, python, numpy)

**Comment by Chengwei94** [comment]
> @Chengwei94 and @smorabit could you give us more info? We can't reproduce your error. In particular we need: all libraries versions (pytorch, CUDA, scanpy, squidpy, python, numpy)

logging==0.5.1.2
anndata==0.7.6
matplotlib==3.4.2
numpy==1.20.3
pandas==1.3.1
scanpy==1.8.1
skimage==0.18.3
squidpy==1.1.1
tangram==1.0.0

Dataset:https://www.10xgenomics.com/resources/datasets/human-breast-cancer-block-a-section-1-1-standard-1-1-0

Code:
``
import logging
import os
import pathlib
import random

import anndata as ad
import matplotlib as mpl
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import scanpy as sc
import skimage
import squidpy as sq
import tangram as tg
from anndata import AnnData
from matplotlib.pyplot import rc_context

# import cell2location
%matplotlib inline

sc.logging.print_header()

ad_st = sc.read_visium(
    "../../../../10X_Data/10x_Genomics_Visium/V1_Breast_Cancer_Block_A_Section_1/"
)
adata_st = sc.pp.subsample(ad_st, fraction=0.7, copy=True)
adata_sc_tmp = ad_st[~(ad_st.obs_names.isin(adata_st.obs_names))]

# Drop adata_sc spatial niformation
adata_sc = AnnData(X=adata_sc.X)
adata_sc.var_names = adata_sc_tmp.var_names
adata_sc.obs_names = adata_sc_tmp.obs_names

# Split into clusters
sc.pp.normalize_total(adata_sc, target_sum=1e4)
sc.pp.log1p(adata_sc)
sc.tl.pca(adata_sc, svd_solver="arpack")
sc.pp.neighbors(adata_sc, n_neighbors=10, n_pcs=40)
sc.tl.leiden(adata_sc)
sc.tl.rank_genes_groups(adata_sc, "leiden", method="t-test")

# Find markers
markers_df = pd.DataFrame(adata_sc.uns["rank_genes_groups"]["names"]).iloc[0:100, :]
markers = list(np.unique(markers_df.melt().value.values))
tg.pp_adatas(adata_sc, adata_st, genes=markers)

ad_map = tg.map_cells_to_space(adata_sc, adata_st,
    mode="cells",
#     mode="clusters",
#     cluster_label='cell_subclass',  # .obs field w cell types
    density_prior='rna_count_based',
    num_epochs=500,
    # device="cuda:0",
    device='cpu',
)

``
Error:

INFO:root:Allocate tensors for mapping.
INFO:root:Begin training with 820 genes and rna_count_based density_prior in cells mode...
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-38-4811e7d5aecd> in <module>
----> 1 ad_map = tg.map_cells_to_space(adata_sc, adata_st,
      2     mode="cells",
      3 #     mode="clusters",
      4 #     cluster_label='cell_subclass',  # .obs field w cell types
      5     density_prior='rna_count_based',

~/miniconda3/envs/singlecell/lib/python3.8/site-packages/tangram/mapping_utils.py in map_cells_to_space(adata_sc, adata_sp, cv_train_genes, cluster_label, mode, device, learning_rate, num_epochs, scale, lambda_d, lambda_g1, lambda_g2, lambda_r, lambda_count, lambda_f_reg, target_count, random_state, verbose, density_prior)
    310             )
    311         )
--> 312         mapper = mo.Mapper(
    313             S=S, G=G, d=d, device=device, random_state=random_state, **hyperparameters,
    314         )

~/miniconda3/envs/singlecell/lib/python3.8/site-packages/tangram/mapping_optimizer.py in __init__(self, S, G, d, d_source, lambda_g1, lambda_d, lambda_g2, lambda_r, device, adata_map, random_state)
     59         self.target_density_enabled = d is not None
     60         if self.target_density_enabled:
---> 61             self.d = torch.tensor(d, device=device, dtype=torch.float32)
     62 
     63         self.source_density_enabled = d_source is not None

ValueError: too many dimensions 'matrix'

**Comment by ziqlu0722** [comment]
> > @Chengwei94 and @smorabit could you give us more info? We can't reproduce your error. In particular we need: all libraries versions (pytorch, CUDA, scanpy, squidpy, python, numpy)
> 
> logging==0.5.1.2
> anndata==0.7.6
> matplotlib==3.4.2
> numpy==1.20.3
> pandas==1.3.1
> scanpy==1.8.1
> skimage==0.18.3
> squidpy==1.1.1
> tangram==1.0.0
> 
> Dataset:https://www.10xgenomics.com/resources/datasets/human-breast-cancer-block-a-section-1-1-standard-1-1-0
> 
> Code:
> 
> ```
> import logging
> import os
> import pathlib
> import random
> import anndata as ad
> import matplotlib as mpl
> import matplotlib.pyplot as plt
> import numpy as np
> import pandas as pd
> import scanpy as sc
> import skimage
> import squidpy as sq
> import tangram as tg
> from anndata import AnnData
> from matplotlib.pyplot import rc_context
> %matplotlib inline
> ```
> 
> ```
> ad_st = sc.read_visium(
>     "../../../../10X_Data/10x_Genomics_Visium/V1_Breast_Cancer_Block_A_Section_1/"
> )
> adata_st = sc.pp.subsample(ad_st, fraction=0.7, copy=True)
> adata_sc_tmp = ad_st[~(ad_st.obs_names.isin(adata_st.obs_names))]
> 
> ### Drop adata_sc spatial information
> adata_sc = AnnData(X=adata_sc.X)
> adata_sc.var_names = adata_sc_tmp.var_names
> adata_sc.obs_names = adata_sc_tmp.obs_names
> 
> # Split into clusters
> sc.pp.normalize_total(adata_sc, target_sum=1e4)
> sc.pp.log1p(adata_sc)
> sc.tl.pca(adata_sc, svd_solver="arpack")
> sc.pp.neighbors(adata_sc, n_neighbors=10, n_pcs=40)
> sc.tl.leiden(adata_sc)
> sc.tl.rank_genes_groups(adata_sc, "leiden", method="t-test")
> 
> # Find markers
> markers_df = pd.DataFrame(adata_sc.uns["rank_genes_groups"]["names"]).iloc[0:100, :]
> markers = list(np.unique(markers_df.melt().value.values))
> tg.pp_adatas(adata_sc, adata_st, genes=markers)
> 
> ad_map = tg.map_cells_to_space(adata_sc, adata_st,
>     mode="cells",
> #     mode="clusters",
> #     cluster_label='cell_subclass',  # .obs field w cell types
>     density_prior='rna_count_based',
>     num_epochs=500,
>     # device="cuda:0",
>     device='cpu',
> )
> ```
> 
> ### Error
> INFO:root:Allocate tensors for mapping.
> INFO:root:Begin training with 820 genes and rna_count_based density_prior in cells mode...
> ValueError Traceback (most recent call last)
> in
> ----> 1 ad_map = tg.map_cells_to_space(adata_sc, adata_st,
> 2 mode="cells",
> 3 # mode="clusters",
> 4 # cluster_label='cell_subclass', # .obs field w cell types
> 5 density_prior='rna_count_based',
> 
> ~/miniconda3/envs/singlecell/lib/python3.8/site-packages/tangram/mapping_utils.py in map_cells_to_space(adata_sc, adata_sp, cv_train_genes, cluster_label, mode, device, learning_rate, num_epochs, scale, lambda_d, lambda_g1, lambda_g2, lambda_r, lambda_count, lambda_f_reg, target_count, random_state, verbose, density_prior)
> 310 )
> 311 )
> --> 312 mapper = mo.Mapper(
> 313 S=S, G=G, d=d, device=device, random_state=random_state, **hyperparameters,
> 314 )
> 
> ~/miniconda3/envs/singlecell/lib/python3.8/site-packages/tangram/mapping_optimizer.py in **init**(self, S, G, d, d_source, lambda_g1, lambda_d, lambda_g2, lambda_r, device, adata_map, random_state)
> 59 self.target_density_enabled = d is not None
> 60 if self.target_density_enabled:
> ---> 61 self.d = torch.tensor(d, device=device, dtype=torch.float32)
> 62
> 63 self.source_density_enabled = d_source is not None
> 
> ValueError: too many dimensions 'matrix'
> 
> ```
> adata_st.obs
> ```
> 
> ValueError Traceback (most recent call last)
> ~/miniconda3/envs/singlecell/lib/python3.8/site-packages/IPython/core/formatters.py in **call**(self, obj)
> 700 type_pprinters=self.type_printers,
> 701 deferred_pprinters=self.deferred_printers)
> --> 702 printer.pretty(obj)
> 703 printer.flush()
> 704 return stream.getvalue()
> 
> ~/miniconda3/envs/singlecell/lib/python3.8/site-packages/IPython/lib/pretty.py in pretty(self, obj)
> 392 if cls is not object
> 393 and callable(cls.**dict**.get('**repr**')):
> --> 394 return _repr_pprint(obj, self, cycle)
> 395
> 396 return _default_pprint(obj, self, cycle)
> 
> ~/miniconda3/envs/singlecell/lib/python3.8/site-packages/IPython/lib/pretty.py in _repr_pprint(obj, p, cycle) 698 """A pprint that just redirects to the normal repr function.""" 699 # Find newlines and replace them with p.break_()
> --> 700 output = repr(obj)
> 701 lines = output.splitlines()
> 702 with p.group():
> 
> ~/miniconda3/envs/singlecell/lib/python3.8/site-packages/pandas/core/frame.py in **repr**(self)
> 993 else:
> 994 width = None
> --> 995 self.to_string(
> 996 buf=buf,
> 997 max_rows=max_rows,
> 
> ~/miniconda3/envs/singlecell/lib/python3.8/site-packages/pandas/core/frame.py in to_string(self, buf, columns, col_space, header, index, na_rep, formatters, float_format, sparsify, index_names, justify, max_rows, min_rows, max_cols, show_dimensions, decimal, line_width, max_colwidth, encoding)
> 1129 decimal=decimal,
> 1130 )
> -> 1131 return fmt.DataFrameRenderer(formatter).to_string(
> 1132 buf=buf,
> 1133 encoding=encoding,
> 
> ~/miniconda3/envs/singlecell/lib/python3.8/site-packages/pandas/io/formats/format.py in to_string(self, buf, encoding, line_width)
> 1051
> 1052 string_formatter = StringFormatter(self.fmt, line_width=line_width)
> -> 1053 string = string_formatter.to_string()
> 1054 return save_to_buffer(string, buf=buf, encoding=encoding)
> 1055
> 
> ~/miniconda3/envs/singlecell/lib/python3.8/site-packages/pandas/io/formats/string.py in to_string(self)
> 23
> 24 def to_string(self) -> str:
> ---> 25 text = self._get_string_representation()
> 26 if self.fmt.should_show_dimensions:
> 27 text = "".join([text, self.fmt.dimensions_info])
> 
> ~/miniconda3/envs/singlecell/lib/python3.8/site-packages/pandas/io/formats/string.py in _get_string_representation(self)
> 38 return self._empty_info_line
> 39
> ---> 40 strcols = self._get_strcols()
> 41
> 42 if self.line_width is None:
> 
> ~/miniconda3/envs/singlecell/lib/python3.8/site-packages/pandas/io/formats/string.py in _get_strcols(self)
> 29
> 30 def _get_strcols(self) -> list[list[str]]:
> ---> 31 strcols = self.fmt.get_strcols()
> 32 if self.fmt.is_truncated:
> 33 strcols = self._insert_dot_separators(strcols)
> 
> ~/miniconda3/envs/singlecell/lib/python3.8/site-packages/pandas/io/formats/format.py in get_strcols(self)
> 538 Render a DataFrame to a list of columns (as lists of strings).
> 539 """
> --> 540 strcols = self._get_strcols_without_index()
> 541
> 542 if self.index:
> 
> ~/miniconda3/envs/singlecell/lib/python3.8/site-packages/pandas/io/formats/format.py in _get_strcols_without_index(self)
> 802 int(self.col_space.get(c, 0)), *(self.adj.len(x) for x in cheader)
> 803 )
> --> 804 fmt_values = self.format_col(i)
> 805 fmt_values = _make_fixed_width(
> 806 fmt_values, self.justify, minimum=header_colwidth, adj=self.adj
> 
> ~/miniconda3/envs/singlecell/lib/python3.8/site-packages/pandas/io/formats/format.py in format_col(self, i)
> 816 frame = self.tr_frame
> 817 formatter = self._get_formatter(i)
> --> 818 return format_array(
> 819 frame.iloc[:, i]._values,
> 820 formatter,
> 
> ~/miniconda3/envs/singlecell/lib/python3.8/site-packages/pandas/io/formats/format.py in format_array(values, formatter, float_format, na_rep, digits, space, justify, decimal, leading_space, quoting)
> 1238 )
> 1239
> -> 1240 return fmt_obj.get_result()
> 1241
> 1242
> 
> ~/miniconda3/envs/singlecell/lib/python3.8/site-packages/pandas/io/formats/format.py in get_result(self)
> 1269
> 1270 def get_result(self) -> list[str]:
> -> 1271 fmt_values = self._format_strings()
> 1272 return _make_fixed_width(fmt_values, self.justify)
> 1273
> 
> ~/miniconda3/envs/singlecell/lib/python3.8/site-packages/pandas/io/formats/format.py in _format_strings(self)
> 1516
> 1517 def _format_strings(self) -> list[str]:
> -> 1518 return list(self.get_result_as_array())
> 1519
> 1520
> 
> ~/miniconda3/envs/singlecell/lib/python3.8/site-packages/pandas/io/formats/format.py in get_result_as_array(self)
> 1480 float_format = lambda value: self.float_format % value
> 1481
> -> 1482 formatted_values = format_values_with(float_format)
> 1483
> 1484 if not self.fixed_width:
> 
> ~/miniconda3/envs/singlecell/lib/python3.8/site-packages/pandas/io/formats/format.py in format_values_with(float_format)
> 1454 values = self.values
> 1455 is_complex = is_complex_dtype(values)
> -> 1456 values = format_with_na_rep(values, formatter, na_rep)
> 1457
> 1458 if self.fixed_width:
> 
> ~/miniconda3/envs/singlecell/lib/python3.8/site-packages/pandas/io/formats/format.py in format_with_na_rep(values, formatter, na_rep)
> 1425 mask = isna(values)
> 1426 formatted = np.array(
> -> 1427 [
> 1428 formatter(val) if not m else na_rep
> 1429 for val, m in zip(values.ravel(), mask.ravel())
> 
> ~/miniconda3/envs/singlecell/lib/python3.8/site-packages/pandas/io/formats/format.py in (.0)
> 1426 formatted = np.array(
> 1427 [
> -> 1428 formatter(val) if not m else na_rep
> 1429 for val, m in zip(values.ravel(), mask.ravel())
> 1430 ]
> 
> ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
> 
> ### Another issue (Just an inconvenience):
> tg.plot_cell_annotation is hardcoded to use x and y as spatial location. However, when loading any visium dataset from spaceranger, the location names are array_rol and array_col.
> 
> ```
> ad_st = sc.read_visium(
>     "../../../../10X_Data/10x_Genomics_Visium/V1_Breast_Cancer_Block_A_Section_1/"
> )
> ad_st
> ```
> 
> AnnData object with n_obs × n_vars = 3798 × 36601
> obs: 'in_tissue', 'array_row', 'array_col'
> var: 'gene_ids', 'feature_types', 'genome'
> uns: 'spatial'
> obsm: 'spatial'

Hi @Chengwei94, currently tangram 1.0.0 is not compatible with Squidpy 1.1.1 before fix of this [issue](https://github.com/theislab/squidpy/issues/408). Could you please try to set up an environment according to [this yml file](https://github.com/broadinstitute/Tangram/blob/master/environment.yml), especially numpy==1.20, pandas==1.2.0, and squidpy 1.1.0?

**Comment by gaddamshreya** [comment]
Hi @Chengwei94 and @smorabit, Tangram now works with the latest version of squidpy and the mentioned issue is now resolved!

**Comment by gaddamshreya** [comment]
Issue is resolved in versions 1.0.1 and newer.

**Comment by Shikari666** [comment]
Tangram package says ---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
[<ipython-input-47-8fbb2a1e63b3>](https://localhost:8080/#) in <cell line: 1>()
----> 1 ad_map = tg.map_cells_to_space(
      2     adata_sc,
      3     adata_st,
      4     mode="constrained",
      5     target_count=adata_st.obs.cell_count.sum(),

AttributeError: module 'tangram' has no attribute 'map_cells_to_space' when trying to run a example snippet code ad_map = tg.map_cells_to_space(
    adata_sc,
    adata_st,
    mode="constrained",
    target_count=adata_st.obs.cell_count.sum(),
    density_prior=np.array(adata_st.obs.cell_count) / adata_st.obs.cell_count.sum(),
    num_epochs=1000,
    device="cpu",
)
